src/Hl7.Fhir.ElementModel/ElementNode.cs

```diff
@@ -0,0 +1,220 @@
+            IElementDefinitionSummary definition = null;
+
+            // Should we throw if type is not found?
+            if (sd != null)
+                definition = new TypeRootDefinitionSummary(sd);
+
+            return new ElementNode(name ?? type, null, type, definition);
+        }
+
+        public static ElementNode FromElement(ITypedElement node, bool recursive = true, IEnumerable<Type> annotationsToCopy = null)
+            => buildNode(node, recursive, annotationsToCopy, null);
+
+        private static ElementNode buildNode(ITypedElement node, bool recursive, IEnumerable<Type> annotationsToCopy, ElementNode parent)
+        {
+            var me = new ElementNode(node.Name, node.Value, node.InstanceType, node.Definition)
+            {
+                Parent = parent
+            };
+
+            foreach (var t in annotationsToCopy ?? Enumerable.Empty<Type>())
+                foreach (var ann in node.Annotations(t))
+                    me.AddAnnotation(ann);
+
+            if (recursive)
+                me.ChildrenInternal.AddRange(node.Children().Select(c => buildNode(c, recursive: true, annotationsToCopy: annotationsToCopy, me)));
+
+            return me;
+        }
+
+        public ElementNode Clone()
+        {
+            var copy = new ElementNode(Name, Value, InstanceType, Definition)
+            {
+                Parent = Parent,
+                ChildrenInternal = ChildrenInternal
+            };
+
+            if (HasAnnotations)
+                copy.AnnotationsInternal.AddRange(AnnotationsInternal);
+
+            return copy;
+        }
+
+        public IElementDefinitionSummary Definition { get; private set; }
+
+        public string InstanceType { get; private set; }
+
+        public object Value { get; set; }
+
+        public IEnumerable<object> Annotations(Type type)
+        {
+            return (type == typeof(ElementNode) || type == typeof(ITypedElement))
+                ? (new[] { this })
+                : AnnotationsInternal.OfType(type);
+        }
+
+        public string Location
+        {
+            get
+            {
+                if (Parent != null)
+                {
+                    //TODO: Slow - but since we'll change the use of this property to informational 
+                    //(i.e. for error messages), it may not be necessary to improve it.
+                    var basePath = Parent.Location;
+
+                    if (Definition?.IsCollection == false)
+                        return $"{basePath}.{Name}";
+                    else
+                    {
+                        var myIndex = Parent.ChildrenInternal.Where(c => c.Name == Name).ToList().IndexOf(this);
+                        return $"{basePath}.{Name}[{myIndex}]";
+                    }
+                    
+                }
+                else
+                    return Name;
+            }
+        }
+
+    }
+
+
+    public class DomNodeList<T> : IEnumerable<T> where T:DomNode<T>
```

Comment:
Do you want this class public?

Reply:
We need to, it's used as the return type of the indexer operator and so is returned to the caller.